transformers/struct.go

```diff
@@ -233,3 +239,99 @@ func TransformWithStruct(st any, opts ...StructTransformerOption) schema.Transfo
 		return nil
 	}
 }
+
+func (t *structTransformer) getColumnType(field reflect.StructField) (arrow.DataType, error) {
+	columnType, err := t.typeTransformer(field)
+	if err != nil {
+		return nil, fmt.Errorf("failed to transform type for field %s: %w", field.Name, err)
+	}
+
+	if columnType == nil {
+		columnType, err = DefaultTypeTransformer(field)
+		if err != nil {
+			return nil, fmt.Errorf("failed to transform type for field %s: %w", field.Name, err)
+		}
+	}
+	return columnType, nil
+}
+
+func structSchemaToJSON(s any) string {
+	b := new(bytes.Buffer)
+	encoder := json.NewEncoder(b)
+	encoder.SetEscapeHTML(false)
+	_ = encoder.Encode(s)
+	return strings.TrimSpace(b.String())
+}
+
+func normalizePointer(field reflect.StructField) reflect.Value {
+	if field.Type.Kind() == reflect.Ptr {
+		return reflect.New(field.Type.Elem())
+	}
+	return reflect.New(field.Type)
+}
+
+func (t *structTransformer) fieldToJSONSchema(field reflect.StructField, depth int) any {
+	transformInput := normalizePointer(field)
+	switch transformInput.Elem().Kind() {
+	case reflect.Struct:
+		fieldsMap := make(map[string]any)
+		fieldType := transformInput.Elem().Type()
+		for i := 0; i < fieldType.NumField(); i++ {
+			name, err := t.nameTransformer(fieldType.Field(i))
+			if err != nil {
+				continue
+			}
+			columnType, err := t.getColumnType(fieldType.Field(i))
+			if err != nil {
+				continue
+			}
+			if columnType == nil {
+				fieldsMap[name] = "any"
+				continue
+			}
+			if columnType == types.ExtensionTypes.JSON && depth < maxJSONTypeSchemaDepth {
+				fieldsMap[name] = t.fieldToJSONSchema(fieldType.Field(i), depth+1)
+				continue
+			}
+			if arrow.IsListLike(columnType.ID()) {
+				fieldsMap[name] = []any{columnType.(*arrow.ListType).Elem().String()}
```

Comment:
Can this typecast ever be `!ok`? I guess I'm asking if a `columnType` that is list like cannot be cast to `*ListType`.

According to godoc:

>// IsListLike returns true for List, LargeList, FixedSizeList, and Map

I wonder if `Map` is an interesting edge case for `.Elem().String()`?

Reply:
So we're the ones creating the list type here https://github.com/cloudquery/plugin-sdk/blob/7955c5358f1360bb0236fc223b490685c7ea5c31/transformers/type.go#L56 and we don't create other types.
However, someone might override the default transformer. I made the condition safer in https://github.com/cloudquery/plugin-sdk/pull/1796/commits/a59d1f7b87e9f7206eeabad80a4933c07a3ac24d